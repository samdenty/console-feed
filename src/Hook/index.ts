import {
  HookedConsole,
  Callback,
  Storage,
  Methods as ConsoleMethods,
  Message,
} from '../definitions/Console'
import Methods from '../definitions/Methods'

import Parse from './parse'
import Unhook from '../Unhook'
import { Encode } from '../Transform'
// import Construct from './construct'

export interface HookOptions {
  encode?: boolean
  async?: boolean
}

const optionsDefault: HookOptions = { encode: true, async: true }

function runImmediately(f: () => void): void {
  f()
}

/**
 * Hook a console constructor and forward messages to a callback
 * @argument console The Console constructor to Hook
 * @argument callback The callback to be called once a message is logged
 */
export default function Hook(
  console: Console,
  callback: Callback,
  optionsIn: boolean | HookOptions = true
) {
  const options: HookOptions = (() => {
    // Support old call style, where third argument is just `encode`
    if (typeof optionsIn === 'boolean') {
      optionsIn = { encode: optionsIn }
    }
    // Set defaults
    optionsIn = Object.assign({}, optionsDefault, optionsIn)
    return optionsIn
  })()

  const TargetConsole = console as HookedConsole
  const Storage: Storage = {
    pointers: {},
    src: {
      npm: 'https://npmjs.com/package/console-feed',
      github: 'https://github.com/samdenty99/console-feed',
    },
  }

  // Override console methods
  for (let method of Methods) {
    const NativeMethod = TargetConsole[method]

    // Override
    TargetConsole[method] = function () {
      // Pass back to native method
      NativeMethod.apply(this, arguments)

      // Parse arguments and send to transport
      const args = [].slice.call(arguments)

      // setTimeout to prevent lag, unless disabled
      const maybeSetTimeout = options.async ? setTimeout : runImmediately
      maybeSetTimeout(() => {
        const parsed = Parse(method as ConsoleMethods, args)
        if (parsed) {
          let encoded: Message = parsed as Message
          if (options.encode) {
            encoded = Encode(parsed) as Message
          }
          callback(encoded, parsed)
        }
      })
    }

    // Store native methods
    Storage.pointers[method] = NativeMethod
  }

  TargetConsole.feed = Storage

  return TargetConsole
}
