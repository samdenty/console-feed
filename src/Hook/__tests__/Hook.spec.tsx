import Hook from '..'
import console from './console'
import Log from './Log'
import { Decode } from '../..'

it('hooks the console', () => {
  Hook(console, (log) => {
    console.logs.push(log)
  })
  expect(console.feed).toBeTruthy()
})

it('forwards log events', async () => {
  const result = await Log('log', 'test')
  expect(result).toBeTruthy()
})

it('decodes messages', () => {
  const decoded = Decode(console.logs[0])
  expect(decoded.method).toEqual('log')
  expect(decoded.data).toMatchSnapshot()
})

it('correctly encodes a HTMLElement', async () => {
  const result = await Log('warn', document.documentElement)
  expect(result).toBeTruthy()

  const decoded = Decode(result)
  expect(decoded.method).toEqual('warn')
  expect(decoded.data).toMatchSnapshot()
})

it('correctly encodes Functions', async () => {
  // prettier-ignore
  const result = await Log('error', function myFunc() { /* body */ })

  const decoded = Decode(result)
  expect(decoded.method).toEqual('error')
  expect(decoded.data).toMatchSnapshot()
})

it('correctly encodes nested values', async () => {
  const input = {
    function: function myFunc() {},
    document: document.documentElement,
    nested: [[[new Promise(() => {})]]],
    recursive: null,
  }
  input.recursive = input

  const result = await Log('debug', input)

  const decoded = Decode(result)
  expect(decoded.method).toEqual('debug')
  expect(decoded.data).toMatchSnapshot()
})

it('disables encoding with a flag', async () => {
  Hook(
    console,
    (log) => {
      console.logs.push(log)
    },
    { encode: false }
  )
  const input = {
    function: function myFunc() {},
    document: document.documentElement,
    nested: [[[new Promise(() => {})]]],
    recursive: null,
  }
  input.recursive = input

  const result: any = await Log('debug', input)

  expect(result.data).toMatchSnapshot()
})

it('disables encoding with a flag (old-style)', async () => {
  Hook(
    console,
    (log) => {
      console.logs.push(log)
    },
    false
  )
  const input = {
    function: function myFunc() {},
    document: document.documentElement,
    nested: [[[new Promise(() => {})]]],
    recursive: null,
  }
  input.recursive = input

  const result: any = await Log('debug', input)

  expect(result.data).toMatchSnapshot()
})

it('disables async with a flag', async () => {
  Hook(
    console,
    (log) => {
      console.logs.push(log)
    },
    { async: false }
  )

  let arr = []
  console.log(arr)
  arr.push(100)
  console.log(arr)

  const decoded1 = Decode(console.logs[console.logs.length - 2])
  const decoded2 = Decode(console.logs[console.logs.length - 1])
  expect(decoded1.method).toEqual('log')
  expect(decoded1.data).toEqual([[]])
  expect(decoded2.method).toEqual('log')
  expect(decoded2.data).toEqual([[100]])
})
