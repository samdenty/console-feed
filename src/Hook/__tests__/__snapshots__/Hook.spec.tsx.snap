// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`correctly encodes Functions 1`] = `
Array [
  [Function],
]
`;

exports[`correctly encodes a HTMLElement 1`] = `
Array [
  <html>
    <head />
    <body />
  </html>,
]
`;

exports[`correctly encodes nested values 1`] = `
Array [
  Object {
    "document": <html>
      <head />
      <body />
    </html>,
    "function": [Function],
    "nested": Array [
      Array [
        Array [
          Object {
            "constructor": Object {
              "name": "Promise",
            },
          },
        ],
      ],
    ],
    "recursive": [Circular],
  },
]
`;

exports[`decodes messages 1`] = `
Array [
  "test",
]
`;

exports[`disables encoding with a flag 1`] = `
Array [
  Object {
    "document": <html>
      <head />
      <body />
    </html>,
    "function": [Function],
    "nested": Array [
      Array [
        Array [
          Promise {},
        ],
      ],
    ],
    "recursive": [Circular],
  },
]
`;
